msal/src/main/java/com/microsoft/identity/client/statemachine/states/BaseStates.kt

```diff
@@ -0,0 +1,289 @@
+import com.microsoft.identity.client.internal.CommandParametersAdapter
+import com.microsoft.identity.client.statemachine.results.SignOutResult
+import com.microsoft.identity.common.internal.commands.AcquireTokenNoFixedScopesCommand
+import com.microsoft.identity.common.internal.commands.RemoveCurrentAccountCommand
+import com.microsoft.identity.common.internal.controllers.LocalMSALController
+import com.microsoft.identity.common.internal.controllers.NativeAuthMsalController
+import com.microsoft.identity.common.java.commands.CommandCallback
+import com.microsoft.identity.common.java.controllers.CommandDispatcher
+import com.microsoft.identity.common.java.controllers.ExceptionAdapter
+import com.microsoft.identity.common.java.dto.AccountRecord
+import com.microsoft.identity.common.java.eststelemetry.PublicApiId
+import com.microsoft.identity.common.java.exception.BaseException
+import com.microsoft.identity.common.java.exception.ServiceException
+import com.microsoft.identity.common.java.logging.LogSession
+import com.microsoft.identity.common.java.logging.Logger
+import com.microsoft.identity.common.java.logging.Logger.LogLevel
+import com.microsoft.identity.common.java.result.ILocalAuthenticationResult
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import java.io.Serializable
+
+sealed interface State
+
+/**
+ * BaseState is the base class for various states in the Native Auth state machine.
+ */
+abstract class BaseState(internal open val flowToken: String?)
+
+/**
+ *  AcccountResult returned as part of a successful completion of sign in flow [com.microsoft.identity.client.statemachine.results.SignInResult.Complete].
+ */
+class AccountResult private constructor(
+    private val account: IAccount,
+    private val config: NativeAuthPublicClientApplicationConfiguration
+) : Serializable {
+
+    companion object {
+
+        private val TAG = NativeAuthPublicClientApplication::class.java.simpleName
+
+        fun createFromAuthenticationResult(
+            authenticationResult: IAuthenticationResult,
+            config: NativeAuthPublicClientApplicationConfiguration
+        ): AccountResult {
+            return AccountResult(
+                account = authenticationResult.account,
+                config = config
+            )
+        }
+
+        fun createFromAccountResult(
+            account: IAccount,
+            config: NativeAuthPublicClientApplicationConfiguration
+        ): AccountResult {
+            return AccountResult(
+                account = account,
+                config = config
+            )
+        }
+    }
+
+    interface SignOutCallback : Callback<SignOutResult>
+
+    /**
+     * Remove the current account from the cache; callback variant.
+     *
+     * @param callback [com.microsoft.identity.client.statemachine.states.AccountResult.SignOutCallback] to receive the result on.
+     */
+    fun signOut(callback: SignOutCallback) {
```

Comment:
So signOut is a method on a base state..why?

Reply:
Does this mean that you can sign out from any given state? What if the user was never even signed in to begin with?

Reply:
If the user does not have a signed in account then we will throw MsalClientException.NO_CURRENT_ACCOUNT error.

Reply:
Ok, to me it sounds like that `signOut` should only be accessible from a signedInState... is there a reason it has to stay on a base state?

Reply:
@shahzaibj  this method (as well as `getAccessToken()`, and other methods) are only available in the `AccountResult` state, which is practically the same as the `SignInState` you mention. They're not part of `BaseState`, this is just an interface.

Reply:
The reason we've called it `AccountResult` rather than `SignedInResult` is because "account" makes more sense to us in this context: on an account object you can retrieve tokens, sign out (the act of signing out an account), etc. This was aligned with our PM.

Reply:
I have moved the `AccountResult` to its own file so its clear that BaseStates interface doesn't have the `signOut` and  `getAccessToken` method.